### ISSUE TITLE
Problems caused by isinstance(list[int], type) returning True

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#89828]
BPO | [45665](https://bugs.python.org/issue45665)
--- | :---
Nosy | @gvanrossum, @serhiy-storchaka, @wyfo, @Fidget-Spinner, @AlexWaygood, @martinitus
Dependencies | <li>bpo-40296: help(list[int]) fails</li><li>bpo-45438: inspect not capturing type annotations created by __class_getitem__</li><li>bpo-45662: Incorrect repr of InitVar of a type alias</li><li>bpo-45663: is_dataclass() does not work for dataclasses which are subclasses of types.GenericAlias</li><li>bpo-45664: resolve_bases() and new_class() do not work with type alias of a built-in type</li><li>bpo-46032: functools&#x27; singledispatch does not support GenericAlias</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-10-29.08:27:43.950>
labels = ['type-bug', 'library', '3.9', '3.10', '3.11']
title = 'Problems caused by isinstance(list[int], type) returning True'
updated_at = <Date 2022-01-06.16:58:47.655>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-06.16:58:47.655>
actor = 'AlexWaygood'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-10-29.08:27:43.950>
creator = 'serhiy.storchaka'
dependencies = ['40296', '45438', '45662', '45663', '45664', '46032']
files = []
hgrepos = []
issue_num = 45665
keywords = []
message_count = 17.0
messages = ['405290', '405291', '405427', '405429', '408207', '409167', '409233', '409679', '409691', '409694', '409697', '409701', '409702', '409703', '409711', '409792', '409881']
nosy_count = 6.0
nosy_names = ['gvanrossum', 'serhiy.storchaka', 'joperez', 'kj', 'AlexWaygood', 'martinitus']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue45665'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

- [x] #84476
- [x] #89601
- [x] #89825
- [x] #89826
- [x] #89827
- [x] #90190
- [x] #93847

**Comment by serhiy-storchaka** [comment]
The cause what that `list[int].__class__` returned `type` instead of `types.GenericAlias`.

I suppose also that there may be problems caused by relaying other dunder methods. There is a list of exceptions, but it is not complete. `typing._GenericAlias` in general does not relay dunder methods (except `__name__` and `__qualname__`).

**Comment by bswck** [comment]
At this point I'd like to report the following inconsistency between C `PyType_Check(t)` and Python's `isinstance(t, type)` that I've noticed. I don't know if it's fixed at the moment, just so that it doesn't get lost:

```py
from abc import ABCMeta  # the C implementation is required for this snippet

class MyABC(metaclass=ABCMeta):
    pass

t = list[int]
isinstance(t, type) and issubclass(t, MyABC)
```
causing the following error:
```
Traceback (most recent call last):
  File "...", line 9, in <module>
    isinstance(t, type) and issubclass(t, MyABC)
  File "/usr/lib/python3.10/abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```
But if we change `_abc` import to `_py_abc`,
 ```py
from _py_abc import ABCMeta  # Python implementation

class MyABC(metaclass=ABCMeta):
    pass

t = list[int]
isinstance(t, type) and issubclass(t, MyABC)
```
the above absolutely does not produce any error.
This is of course clearly a `GenericAlias`-related problem.

**Comment by gvanrossum** [comment]
> At this point I'd like to report the following inconsistency between C `PyType_Check(t)` and Python's `isinstance(t, type)` that I've noticed. I don't know if it's fixed at the moment, just so that it doesn't get lost:
> [...]

What Python version? The most recent version where I can repro that error is 3.10. The fix appears to be in 3.11 and on main (3.12 alpha).

**Comment by bswck** [comment]
Oh, right. Yes, it was 3.10, like in the traceback. 
Sorry for not specifying the version and not checking if the bug is not already fixed in versions >3.10.
Also sorry for causing unnecessary chaos. 
Thanks and good luck!

**Comment by serhiy-storchaka** [comment]
There is not much that we can do with this in 3.10. The workaround in the user code is:
```python
isinstance(t, type) and not isinstance(t, types.GenericAlias) and issubclass(t, MyABC)
```
The middle check is not needed in 3.11+.

I think this issue can now be closed. I have not found other sites in the stdlib with user visible differences.

But I have found an interesting difference in the behavior of `issubclass()`, which should rarely affect user code: #101162.
